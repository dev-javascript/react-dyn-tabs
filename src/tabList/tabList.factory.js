import React from 'react';
import {ApiContext, StateContext} from '../utils/context.js';
function TabList(deps, props, ref) {
  const {openTabIDs, selectedTabID} = React.useContext(StateContext);
  const {Tabs, tablistPropsGenerator} = deps(React.useContext(ApiContext), ref, openTabIDs, selectedTabID);
  return <Tabs {...tablistPropsGenerator()} />;
}
export default TabList;
