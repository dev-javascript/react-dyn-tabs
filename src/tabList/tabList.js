import {memo, forwardRef} from 'react';
import TabList from './tabList.factory.js';
import Tabs from '../tabs/tabs.js';
const MemomizedTabList = memo(
  forwardRef(
    TabList.bind(undefined, (api, ref, state) => ({
      Tabs,
      tablistPropsGenerator: () => api.tablistPropsGenerator(ref, state),
    })),
  ),
  () => true,
);
export default MemomizedTabList;
