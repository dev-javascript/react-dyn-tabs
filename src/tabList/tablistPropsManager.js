export default function ({optionsManager: {setting: _setting, options: _op}}) {
  const {tablistRootClass, tablistSectionClass, tablistSliderClass, tablistClass} = _setting,
    result = {
      tablistRootProps: {className: tablistRootClass},
      tablistSectionProps: {className: tablistSectionClass},
      tablistSliderProps: {className: tablistSliderClass + ' ' + tablistClass},
      tablistProps: {className: tablistClass},
    };
  if (_op.isVertical) {
    result.tablistRootProps.className += ' ' + _setting.verticalClass;
    result.tablistProps.className += ' ' + _setting.verticalClass;
    result.tablistSliderProps.className += ' ' + _setting.verticalClass;
  }
  if (_op.direction === 'rtl') {
    result.tablistRootProps.className += ' ' + _setting.rtlClass;
    result.tablistProps.className += ' ' + _setting.rtlClass;
    result.tablistSliderProps.className += ' ' + _setting.rtlClass;
  }
  if (_op.accessibility) {
    result.tablistProps.role = 'tablist';
  }

  return result;
}
