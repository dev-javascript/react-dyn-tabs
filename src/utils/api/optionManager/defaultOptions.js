import React from 'react';
const DefaultOptions = function (DefaulTabInnerComponent = null) {
  this.defaultDirection = 'ltr';
  this._DefaulTabInnerComponent = DefaulTabInnerComponent;
  this.directionsRange = ['ltr', 'rtl'];
};
Object.assign(DefaultOptions.prototype, {
  getOptions: function () {
    return this._getOptions();
  },
  _getOptions: function () {
    const _options = {
      tabs: [],
      selectedTabID: '',
      beforeSelect: function () {
        return true;
      },
      beforeClose: function () {
        return true;
      },
      onOpen: function () {},
      onClose: function () {},
      onFirstSelect: function () {},
      onSelect: function () {},
      onChange: function () {},
      onLoad: function () {},
      onDestroy: function () {},
      onInit: function () {},
      accessibility: true,
      isVertical: false,
      defaultPanelComponent: function defaultPanelComponent() {
        return <div></div>;
      },
      showMoreButtonComponent: null,
    };
    let _direction = this.defaultDirection,
      _tabComponent = this._DefaulTabInnerComponent;
    const that = this;
    Object.defineProperties(_options, {
      direction: {
        get() {
          return _direction;
        },
        set(value) {
          if (that.directionsRange.indexOf(value) === -1)
            throw 'Invalid direction value! it can be eather of "ltr" or "rtl" ';
          _direction = value;
        },
        enumerable: true,
      },
      tabComponent: {
        get() {
          return _tabComponent;
        },
        set(fn) {
          if (fn && typeof fn !== 'function') throw 'tabComponent property must be type of a function.';
          _tabComponent = fn ? fn : that._DefaulTabInnerComponent;
        },
        enumerable: true,
      },
    });
    return _options;
  },
});
export default DefaultOptions;
