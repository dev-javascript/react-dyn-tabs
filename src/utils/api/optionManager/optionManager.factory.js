import React from 'react';
import Helper from '../../helper.js';
const {throwMissingParam: missingParamEr, uuid} = Helper;
export default Helper.module(
  function (getDeps, {options}) {
    const {globalDefaultOptions, internalOptions} = getDeps();
    this._defaultOptions = globalDefaultOptions;
    this.internalOptions = internalOptions;
    this._validateOptions(options);
    this.options = Object.assign({}, this._defaultOptions, options);
    this.setting = {};
    this.initialState = {};
    this.initialTabs = [];
    this._setSetting()._setInitialData();
  },
  {
    getOption: function (optionName) {
      if (optionName === 'tabs') {
        // returned result should be immutable
        let arr = [];
        for (let i = 0, tabs = this.options.tabs, l = tabs.length; i < l; i++) {
          arr.push({...tabs[i]});
        }
        return arr;
      }
      return this.options[optionName];
    },
    setOption: function (name = missingParamEr('setOption'), value = missingParamEr('setOption')) {
      if (['SELECTEDTABID', 'TABS'].indexOf(name.toUpperCase()) >= 0) return this;
      if (Object.prototype.hasOwnProperty.call(this._defaultOptions, name)) this.options[name] = value;
      return this;
    },
    validatePanelComponent: function (tabData) {
      // convert panel element into a  function component.
      if (
        tabData.panelComponent &&
        typeof tabData.panelComponent !== 'function' &&
        React.isValidElement(tabData.panelComponent)
      ) {
        const PanelElement = tabData.panelComponent;
        tabData.panelComponent = function () {
          return PanelElement;
        };
      }
      return this;
    },
    validateObjectiveTabData: function (tabData) {
      if (Object.prototype.toString.call(tabData) !== '[object Object]')
        throw new Error('tabData must be type of Object');
      return this;
    },
    validateTabData: function (tabData) {
      this.validateObjectiveTabData(tabData).validatePanelComponent(tabData);
      tabData = Object.assign(this.setting.getDefaultTabData(), tabData);
      tabData.id = tabData.id + ''; //make sure id is string
      return tabData;
    },
    _validateOptions: function (options) {
      if (Object.prototype.toString.call(options) !== '[object Object]')
        throw 'Invalid argument in "useDynamicTabs" function. Argument must be type of an object';
      return this;
    },
    _setInitialData: function () {
      // set this.initialTabs and this.initialState
      const {selectedTabID, tabs} = this.options,
        openTabIDs = [];
      tabs.forEach((tab) => {
        const newTab = this.validateTabData(tab);
        this.initialTabs.push(newTab);
        openTabIDs.push(newTab.id);
      });
      this.initialState = {
        selectedTabID: selectedTabID + '', //make sure it is type of string
        openTabIDs,
      };
      return this;
    },
    _setSetting: function () {
      this.setting = {
        tabClass: 'rc-dyn-tabs-tab',
        titleClass: 'rc-dyn-tabs-title',
        iconClass: 'rc-dyn-tabs-icon',
        selectedClass: 'rc-dyn-tabs-selected',
        hoverClass: 'rc-dyn-tabs-hover',
        closeClass: 'rc-dyn-tabs-close',
        panelClass: 'rc-dyn-tabs-panel',
        panellistClass: 'rc-dyn-tabs-panellist',
        disableClass: 'rc-dyn-tabs-disable',
        ltrClass: 'rc-dyn-tabs-ltr',
        rtlClass: 'rc-dyn-tabs-rtl',
        verticalClass: 'rc-dyn-tabs-vertical',
        tablistViewClass: 'rc-dyn-tabs-tablist-view',
        tablistContainerClass: 'rc-dyn-tabs-tablist-container',
        tablistOverflowClass: 'rc-dyn-tabs-tablist-overflow',
        tablistClass: 'rc-dyn-tabs-tablist',
        responsiveClass: '',
        showMoreContainerClass: 'rc-dyn-tabs-showmorebutton-container',
        showMoreButtonClass: 'rc-dyn-tabs-showmorebutton',
        showMorePopperClass: 'rc-dyn-tabs-popper',
        panelIdTemplate: (id) => `rc-dyn-tabs-p-${id}`,
        ariaLabelledbyIdTemplate: (id) => `rc-dyn-tabs-l-${id}`,
        getDefaultTabData: () => {
          return {
            title: '',
            tooltip: '',
            panelComponent: this.options.defaultPanelComponent,
            closable: true,
            iconClass: '',
            disable: false,
            lazy: false,
            id: `tab_${uuid()}`,
          };
        },
      };
      return this;
    },
  },
);
