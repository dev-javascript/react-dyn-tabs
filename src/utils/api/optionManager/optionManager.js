import factory from './optionManager.factory.js';
import DefaultTabInnerComponent from '../../../tab/defaulTabInner.js';
import DefaultOptions from './defaultOptions.js';
const getDeps = function () {
  const globalDefaultOptions = new DefaultOptions(DefaultTabInnerComponent).getOptions();
  return {globalDefaultOptions};
};
export default factory.bind(null, getDeps);
