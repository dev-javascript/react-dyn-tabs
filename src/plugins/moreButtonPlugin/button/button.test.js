/* eslint-disable react/prop-types */
import React, { forwardRef } from 'react';
import { unmountComponentAtNode } from 'react-dom';
import ButtonComponent from './button.js';
import PopperComponent from './popper/popper.js';
import renderer from 'react-test-renderer';
import optionManager from '../../../../utils/api/optionManager/optionManager.js';
import { Tabs } from '../../../../tabList/tabList.js';
let container = document.createElement('div');
const Popper = forwardRef(
  PopperComponent.bind(null, {
    createPopper: () => { },
    getPopperMaxHeight: () => 200,
    clk: () => { },
  }),
);
const Button = ButtonComponent.bind(null, () => ({ Popper }));
const opManager = new optionManager({ options: {} });
const st = opManager.setting;
const getApi = (props = {}) =>
  Object.assign({}, { on: () => { }, of: () => { }, getData: () => ({ selectedTabID: '0' }), getOption: () => 'ltr' }, props);

beforeAll(() => {
  document.body.appendChild(container);
});
beforeEach(() => { });
afterEach(() => {
  unmountComponentAtNode(container);
  container.innerHTML = '';
});
afterAll(() => {
  document.body.removeChild(container);
  container = null;
});
describe('Default show more button structure : ', () => {
  test('default', () => {
    const tree = renderer
      .create(
        <div className={st.tabClass + ' ' + st.showMoreContainerClass}>
          <Button
            hiddenTabIDs={'1,2'}
            instance={getApi()}
            TabsComponent={Tabs}
            buttonClassName={st.titleClass + ' ' + st.showMoreButtonClass}
            popperClassName={st.tablistViewClass + ' ' + st.verticalClass + ' ' + st.showMorePopperClass}
          />
        </div>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  test('different hiddenTabIDs', () => {
    const tree = renderer
      .create(
        <div className={st.tabClass + ' ' + st.showMoreContainerClass}>
          <Button
            hiddenTabIDs={'1,3,4'}
            instance={getApi()}
            TabsComponent={Tabs}
            buttonClassName={st.titleClass + ' ' + st.showMoreButtonClass}
            popperClassName={st.tablistViewClass + ' ' + st.verticalClass + ' ' + st.showMorePopperClass}
          />
        </div>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  test('rtl mode', () => {
    const tree = renderer
      .create(
        <div className={st.tabClass + ' ' + st.showMoreContainerClass}>
          <Button
            hiddenTabIDs={'1,2'}
            instance={getApi({ getOption: () => 'rtl' })}
            TabsComponent={Tabs}
            buttonClassName={st.titleClass + ' ' + st.showMoreButtonClass}
            popperClassName={
              st.tablistViewClass + ' ' + st.verticalClass + ' ' + st.rtlClass + ' ' + st.showMorePopperClass
            }
          />
        </div>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
