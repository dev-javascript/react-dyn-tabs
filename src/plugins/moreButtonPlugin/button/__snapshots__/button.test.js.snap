// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DISABLE ACCESIBILITY :  DEFAULT: 1`] = `
<div>
  <div
    className="rc-dyn-tabs-title rc-dyn-tabs-showmorebutton"
    onClick={[Function]}
  >
    <div
      instance={
        {
          "close": [Function],
          "getCopyData": [Function],
          "getCopyPerviousData": [Function],
          "getData": [Function],
          "getOption": [Function],
          "getPreviousData": [Function],
          "getTab": [Function],
          "isOpen": [Function],
          "isSelected": [Function],
          "off": [Function],
          "on": [Function],
          "one": [Function],
          "open": [Function],
          "refresh": [Function],
          "select": [Function],
          "setOption": [Function],
          "setTab": [Function],
          "sort": [Function],
        }
      }
    >
      more button mock icon
    </div>
  </div>
</div>
`;

exports[`Enable Accesibility :  DEFAULT: 1`] = `
<div>
  <div
    aria-controls="rc-dyn-tabs-popupcontainer_0"
    aria-expanded={false}
    aria-haspopup="true"
    className="rc-dyn-tabs-title rc-dyn-tabs-showmorebutton"
    id="rc-dyn-tabs-more-button-0"
    onClick={[Function]}
    role="button"
    tabIndex={0}
  >
    <div
      instance={
        {
          "close": [Function],
          "getCopyData": [Function],
          "getCopyPerviousData": [Function],
          "getData": [Function],
          "getOption": [Function],
          "getPreviousData": [Function],
          "getTab": [Function],
          "isOpen": [Function],
          "isSelected": [Function],
          "off": [Function],
          "on": [Function],
          "one": [Function],
          "open": [Function],
          "refresh": [Function],
          "select": [Function],
          "setOption": [Function],
          "setTab": [Function],
          "sort": [Function],
        }
      }
    >
      more button mock icon
    </div>
  </div>
</div>
`;

exports[`Enable Accesibility :  MULTIPLE BUTTON SHOULD NOT HAVE SAME ID: 1`] = `
<div>
  <div
    aria-controls="rc-dyn-tabs-popupcontainer_1"
    aria-expanded={false}
    aria-haspopup="true"
    className="rc-dyn-tabs-title rc-dyn-tabs-showmorebutton"
    id="rc-dyn-tabs-more-button-1"
    onClick={[Function]}
    role="button"
    tabIndex={0}
  >
    <div
      instance={
        {
          "close": [Function],
          "getCopyData": [Function],
          "getCopyPerviousData": [Function],
          "getData": [Function],
          "getOption": [Function],
          "getPreviousData": [Function],
          "getTab": [Function],
          "isOpen": [Function],
          "isSelected": [Function],
          "off": [Function],
          "on": [Function],
          "one": [Function],
          "open": [Function],
          "refresh": [Function],
          "select": [Function],
          "setOption": [Function],
          "setTab": [Function],
          "sort": [Function],
        }
      }
    >
      more button mock icon
    </div>
  </div>
  <div
    aria-controls="rc-dyn-tabs-popupcontainer_2"
    aria-expanded={false}
    aria-haspopup="true"
    className="rc-dyn-tabs-title rc-dyn-tabs-showmorebutton"
    id="rc-dyn-tabs-more-button-2"
    onClick={[Function]}
    role="button"
    tabIndex={0}
  >
    <div
      instance={
        {
          "close": [Function],
          "getCopyData": [Function],
          "getCopyPerviousData": [Function],
          "getData": [Function],
          "getOption": [Function],
          "getPreviousData": [Function],
          "getTab": [Function],
          "isOpen": [Function],
          "isSelected": [Function],
          "off": [Function],
          "on": [Function],
          "one": [Function],
          "open": [Function],
          "refresh": [Function],
          "select": [Function],
          "setOption": [Function],
          "setTab": [Function],
          "sort": [Function],
        }
      }
    >
      more button mock icon
    </div>
  </div>
</div>
`;

exports[`TABS COMPONENT: ltr: 1`] = `
<ul
  aria-labelledby="rc-dyn-tabs-more-button-8"
  className="rc-dyn-tabs-tablist rc-dyn-tabs-ltr rc-dyn-tabs-vertical"
  id="rc-dyn-tabs-popupcontainer_8"
  role="menu"
>
  <li
    aria-label="tab1"
    className="rc-dyn-tabs-tab rc-dyn-tabs-selected"
    onClick={[Function]}
    role="menuitem"
    tab-id="1"
    tabIndex={0}
    title="tab1"
  >
    <button
      className="rc-dyn-tabs-title"
      role="presentation"
      tab-id="1"
      tabIndex={-1}
    >
      tab1
    </button>
    <span
      className="rc-dyn-tabs-close"
      role="button"
    >
      ×
    </span>
  </li>
  <li
    aria-label="tab2"
    className="rc-dyn-tabs-tab"
    onClick={[Function]}
    role="menuitem"
    tab-id="2"
    tabIndex={-1}
    title="tab2"
  >
    <button
      className="rc-dyn-tabs-title"
      role="presentation"
      tab-id="2"
      tabIndex={-1}
    >
      tab2
    </button>
    <span
      className="rc-dyn-tabs-close"
      role="button"
    >
      ×
    </span>
  </li>
</ul>
`;

exports[`TABS COMPONENT: rtl: 1`] = `
<ul
  aria-labelledby="rc-dyn-tabs-more-button-9"
  className="rc-dyn-tabs-tablist rc-dyn-tabs-rtl rc-dyn-tabs-vertical"
  id="rc-dyn-tabs-popupcontainer_9"
  role="menu"
>
  <li
    aria-label="tab1"
    className="rc-dyn-tabs-tab rc-dyn-tabs-selected"
    onClick={[Function]}
    role="menuitem"
    tab-id="1"
    tabIndex={0}
    title="tab1"
  >
    <button
      className="rc-dyn-tabs-title"
      role="presentation"
      tab-id="1"
      tabIndex={-1}
    >
      tab1
    </button>
    <span
      className="rc-dyn-tabs-close"
      role="button"
    >
      ×
    </span>
  </li>
  <li
    aria-label="tab2"
    className="rc-dyn-tabs-tab"
    onClick={[Function]}
    role="menuitem"
    tab-id="2"
    tabIndex={-1}
    title="tab2"
  >
    <button
      className="rc-dyn-tabs-title"
      role="presentation"
      tab-id="2"
      tabIndex={-1}
    >
      tab2
    </button>
    <span
      className="rc-dyn-tabs-close"
      role="button"
    >
      ×
    </span>
  </li>
</ul>
`;

exports[`TABS COMPONENT: trun off accesibility: 1`] = `
<ul
  className="rc-dyn-tabs-tablist rc-dyn-tabs-ltr rc-dyn-tabs-vertical"
>
  <li
    className="rc-dyn-tabs-tab rc-dyn-tabs-selected"
    onClick={[Function]}
    tab-id="1"
    tabIndex={0}
    title="tab1"
  >
    <button
      className="rc-dyn-tabs-title"
      role="presentation"
      tab-id="1"
      tabIndex={-1}
    >
      tab1
    </button>
    <span
      className="rc-dyn-tabs-close"
    >
      ×
    </span>
  </li>
  <li
    className="rc-dyn-tabs-tab"
    onClick={[Function]}
    tab-id="2"
    tabIndex={-1}
    title="tab2"
  >
    <button
      className="rc-dyn-tabs-title"
      role="presentation"
      tab-id="2"
      tabIndex={-1}
    >
      tab2
    </button>
    <span
      className="rc-dyn-tabs-close"
    >
      ×
    </span>
  </li>
</ul>
`;
